akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala

```diff
@@ -0,0 +1,1577 @@
+          val merged = envelope.merge(pruningCleanupTombstoned(writeEnvelope)).addSeen(selfAddress)
+          setData(key, merged)
+        } else {
+          log.warning("Wrong type for writing [{}], existing type [{}], got [{}]",
+            key, existing.getClass.getName, writeEnvelope.data.getClass.getName)
+        }
+      case None ⇒
+        setData(key, pruningCleanupTombstoned(writeEnvelope).addSeen(selfAddress))
+    }
+
+  def receiveReadRepair(key: String, writeEnvelope: DataEnvelope): Unit = {
+    write(key, writeEnvelope)
+    sender() ! ReadRepairAck
+  }
+
+  def receiveGetKeys(): Unit = {
+    val keys: Set[String] = dataEntries.collect {
+      case (key, (DataEnvelope(data, _), _)) if data != DeletedData ⇒ key
+    }(collection.breakOut)
+    sender() ! GetKeysResult(keys)
+  }
+
+  def receiveDelete(key: String, consistency: WriteConsistency, replyTo: ActorRef): Unit = {
+    // don't use sender() in this method, since it is used from drainUpdateInProgressBuffer
+    getData(key) match {
+      case Some(DataEnvelope(DeletedData, _)) ⇒
+        // already deleted
+        replyTo ! DataDeleted(key)
+      case _ ⇒
+        setData(key, DeletedEnvelope)
+        if (isLocalUpdate(consistency))
+          replyTo ! DeleteSuccess(key)
+        else
+          context.actorOf(WriteAggregator.props(key, DeletedEnvelope, consistency, None, nodes, replyTo))
+    }
+  }
+
+  def setData(key: String, envelope: DataEnvelope): Unit = {
+    // notify subscribers, later
+    if (subscribers.contains(key) && !changed.contains(key)) {
+      val oldDigest = getDigest(key)
+      changed = changed.updated(key, oldDigest)
+    }
+
+    val dig = if (envelope.data == DeletedData) DeletedDigest else LazyDigest
+    dataEntries = dataEntries.updated(key, (envelope, dig))
+  }
+
+  def getDigest(key: String): Digest = {
+    dataEntries.get(key) match {
+      case Some((envelope @ DataEnvelope(_, _), LazyDigest)) ⇒
+        val d = digest(envelope)
+        dataEntries = dataEntries.updated(key, (envelope, d))
+        d
+      case Some((_, digest)) ⇒ digest
+      case None              ⇒ NotFoundDigest
+    }
+  }
+
+  def digest(envelope: DataEnvelope): Digest = {
+    val bytes = serializer.toBinary(envelope)
+    ByteString.fromArray(MessageDigest.getInstance("SHA-1").digest(bytes))
+  }
+
+  def getData(key: String): Option[DataEnvelope] = dataEntries.get(key).map { case (envelope, _) ⇒ envelope }
+
+  def receiveFlushChanges(): Unit = {
+    def notify(key: String, subs: mutable.Set[ActorRef]): Unit = {
+      getData(key) match {
+        case Some(envelope) ⇒
+          val msg = if (envelope.data == DeletedData) DataDeleted(key) else Changed(key, envelope.data)
+          subs.foreach { _ ! msg }
+        case None ⇒
+      }
+    }
+
+    if (subscribers.nonEmpty) {
+      for ((key, oldDigest) ← changed; subs ← subscribers.get(key)) {
+        if (oldDigest != getDigest(key))
```

Comment:
I'm not sure, but I don't think it's needed. I might have been thinking that for changes back and forth resulting in same content should not emit an event, but that is not a valid crdt as far as I can see. I'll remove it.